src/java.base/share/classes/java/lang/Enum.java

```diff
@@ -166,13 +168,29 @@ public final boolean equals(Object other) {
         return this==other;
     }
 
+    /**
+     * The hash code of this enumeration constant.
+     *
+     * @implNote Once initialized, the field value does not change.
+     * Hotspot's identity hash code generation also never returns zero
```

Comment:
It would not break the code functionally if that invariant ever breaks: we would "just" call the (intrinsic) method on zero hash code. That `implNote` only shows that it would not happen with current implementation at all.

Reply:
From that impl note it seemed like it was a big deal for hash code to never return 0 for an object. Could you maybe de-emphasize the importance of that HotSpot behavior in the note?

Reply:
All right, we can change "This allows to treat zero as the marker..." to "This makes zero a convenient marker...", I think?

Reply:
To me, the proposed implementation looks similar to that of String::hashCode before Compact Strings (JEP 254).

The interesting detail here is the fact that unlike String::hashCode, which can legitimately return 0, Object::hashCode cannot, at least in HotSpot. This is great news, because to be performant, java.lang.Enum does not need to be bloated with extra mechanics, such as that boolean `hashIsZero` introduced for java.lang.String in 8221836: Avoid recalculating String.hash when zero.

If the essence of that can be compressed to a few simple sentences and possibly moved to an inline comment in Enum::hashCode, that _might_ be better.